pkg/filesystem/path/windows_parser.go

```diff
@@ -0,0 +1,138 @@
+package path
+
+import (
+	"strings"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+func stripWindowsSeparators(p string) string {
+	for p != "" && (p[0] == '/' || p[0] == '\\') {
+		p = p[1:]
+	}
+	return p
+}
+
+type windowsParser struct {
+	path string
+}
+
+type windowsRelativeParser struct {
+	path string
+}
+
+// NewWindowsParser creates a Parser for Windows paths that can be used in Resolve.
+func NewWindowsParser(path string) (Parser, error) {
+	// Windows paths are generally passed to system calls that
+	// accept C strings. There is no way these can accept null
+	// bytes.
+	if strings.ContainsRune(path, '\x00') {
+		return nil, status.Error(codes.InvalidArgument, "Path contains a null byte")
+	}
+	return &windowsParser{path}, nil
+}
+
+// MustNewWindowsParser is identical to NewWindowsParser, except that it panics
+// upon failure.
+func MustNewWindowsParser(path string) Parser {
+	parser, err := NewWindowsParser(path)
+	if err != nil {
+		panic(err)
+	}
+	return parser
+}
+
+func (p windowsParser) ParseScope(scopeWalker ScopeWalker) (next ComponentWalker, remainder RelativeParser, err error) {
+	// Parse for drive letter.
+	if len(p.path) >= 2 && ((p.path[0] >= 'A' && p.path[0] <= 'Z') ||
+		(p.path[0] >= 'a' && p.path[0] <= 'z')) && p.path[1] == ':' {
+		next, err = scopeWalker.OnDriveLetter(rune(p.path[0] &^ 0x20))
+		if err != nil {
+			return nil, nil, err
+		}
+
+		return next, windowsRelativeParser{stripWindowsSeparators(p.path[2:])}, nil
+	}
+
+	if len(p.path) >= 1 && p.path[0] == '\\' || p.path[0] == '/' {
+		next, err = scopeWalker.OnAbsolute()
+		if err != nil {
+			return nil, nil, err
+		}
+
+		return next, windowsRelativeParser{stripWindowsSeparators(p.path)}, nil
+	}
+
+	next, err = scopeWalker.OnRelative()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return next, windowsRelativeParser{p.path}, nil
+}
+
+func (rp windowsRelativeParser) ParseFirstComponent(componentWalker ComponentWalker, mustBeDirectory bool) (next GotDirectoryOrSymlink, remainder RelativeParser, err error) {
+	var name string
+	terminal := false
+
+	separator := strings.IndexAny(rp.path, "/\\")
+	if separator == -1 {
+		// Path no longer contains a separator. Consume it entirely.
+		terminal = true
+		name = rp.path
+		remainder = nil
+	} else {
+		name = rp.path[:separator]
+		rp.path = stripWindowsSeparators(rp.path[separator:])
+		remainder = windowsRelativeParser{rp.path}
+	}
+
+	switch name {
+	case "", ".":
+		// An explicit "." entry, or an empty component.
+		// Empty components can occur if paths end with
+		// one or more slashes. Treat "foo/" as identical
+		// to "foo/."
+		return GotDirectory{Child: componentWalker}, remainder, nil
+	case "..":
+		// Traverse to the parent directory.
+		parent, err := componentWalker.OnUp()
+		if err != nil {
+			return nil, nil, err
+		}
+		return GotDirectory{Child: parent}, remainder, nil
+	}
+
```

Comment:
Could we add a call to `validateWindowsPathComponent()` here? This ensures that we also don't parse malformed paths.

Reply:
Adding such a check would also allow you to get rid of
```go
	if strings.ContainsRune(path, '\x00') {
		return nil, status.Error(codes.InvalidArgument, "Path contains a null byte")
	}
```
Because `validateWindowsPathComponent()` also does that for you.

Reply:
Thinking slightly ahead: if the `strings.ContainsRune(path, '\x00')` check is removed from `NewWindowsParser()`, then it can no longer return any errors. That also removes the need for having `MustNewWindowsParser()`.